aspnetcore/tutorials/web-api-vsc.md

```diff
@@ -0,0 +1,272 @@
+
+* To keep the tutorial simple, the app doesn’t use a persistent database. Instead, it stores to-do items in an in-memory database. 
+
+### Set up your development environment
+
+Download and install [.NET Core](https://microsoft.com/net/core) and [Visual Studio Code](https://code.visualstudio.com) with the [C# extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.csharp).
+
+### Create the project
+
+From a console, run the following commands:
+
+```console
+mkdir TodoApi
+cd TodoApi
+dotnet new webapi
+```
+
+Open the *TodoApi* folder in Visual Studio Code (VS Code). 
+
+- Select **Yes** to the **Warn** message "Required assets to build and debug are missing from 'TodoApi'. Add them?"
+- Select **Restore** to the **Info** message "There are unsresolved dependencies".
+
+![VS Code with Warn Required assets to build and debug are missing from 'TodoApi'. Add them? Don't ask Again, Not Now, Yes and also Info - there are unresolved dependencies  - Restore - Close](web-api-vsc/_static/vsc_restore.png)
+
+Press ^F5 to build and run the program. In a browser naviage to http://localhost:5000/api/values . The following is displayed:
+
+`["value1","value2"]`
+
+### Add support for Entity Framework Core
+
+Edit the *TodoApi.csproj* file to install the [Entity Framework Core InMemory](https://docs.microsoft.com/en-us/ef/core/providers/in-memory/) database provider. This database provider allows Entity Framework Core to be used with an in-memory database.
+
+[!code-xml[Main](web-api-vsc/sample/TodoApi/TodoApi.csproj?highlight=13-15)]
+
+Run `dotnet restore` to download and install the EF Core InMemory DB provider. You can run `dotnet restore` from the terminal or enter `⌘⇧P` or `Ctrl+Shift+P` in VS Code and then type **.NET**. Select **.NET: Restore Packages**.
+
+
+### Add a model class
+
+A model is an object that represents the data in your application. In this case, the only model is a to-do item.
+
+Add a folder named *Models*. You can put model classes anywhere in your project, but the *Models* folder is used by convention.
+
+Add a `TodoItem` class with the following code:
+
+[!code-csharp[Main](first-web-api/sample/TodoApi/Models/TodoItem.cs)]
+
+* The `[Key]` data annotation denotes the property, `Key`, is a unique identifier.
+* `[DatabaseGenerated` specifies the database will generate the key (rather than the application).
+* `DatabaseGeneratedOption.Identity` specifies the database should generate integer keys when a row is inserted. 
+
+### Create the database context
+
+The *database context* is the main class that coordinates Entity Framework functionality for a given data model. You create this class by deriving from the `Microsoft.EntityFrameworkCore.DbContext` class.
+
+Add a `TodoContext` class in the *Models* folder:
+
+[!code-csharp[Main](first-web-api/sample/TodoApi/Models/TodoContext.cs)]
+
+## Add a repository class
```

Comment:
This approach makes it easier to unit test your controllers. Unit tests should inject a mock or stub version of `ITodoRepository`. That way, the test narrowly targets the controller logic and not the data access layer.